swank/backend.lisp

```diff
@@ -727,6 +727,43 @@ NIL."
                    (values new-form expanded)))))
     (frob form env)))
 
+(defmacro with-collected-macro-forms ((var) instrumented-form &body body)
+  "Collect macro forms by locally binding *MACROEXPAND-HOOK*.
+
+Evaluates INSTRUMENTED-FORM, with *MACROEXPAND-HOOK* bound to a
+function which collects all forms that undergo macro-expansion into a
+list stored in VAR.  Then evaluates BODY, with VAR still bound to the
+list of forms."
+  (assert (symbolp var))
+  (assert (not (null var)))
+  (let ((real-macroexpand-hook (gensym))
+        (instrumented-macroexpand-hook (gensym)))
+    `(let* ((,real-macroexpand-hook *macroexpand-hook*)
+            (,var '())
+            (,instrumented-macroexpand-hook
+             (lambda (macro-function form environment)
+               (let ((result (funcall ,real-macroexpand-hook
+                                      macro-function form environment)))
+                 (unless (eq result form)
+                   (setq ,var (cons form ,var)))
```

Comment:
Minor refactoring suggestion: `(push form ,var)`.

Slightly less minor refactoring suggestion, move non-syntactic bits to a function.

```lisp
(defun call-with-collected-macro-forms (body-fn instrumented-form-fn)
  (let* ((real-macroexpand-hook *macroexpand-hook*)
	 (forms '())
	 (instrumented-macroexpand-hook
	   (lambda (macro-function form environment)
	     (let ((result (funcall real-macroexpand-hook
				    macro-function form environment)))
	       (unless (eq result form)
		 (push form forms))
	       result))))
    (let ((*macroexpand-hook* instrumented-macroexpand-hook))
      (funcall instrumented-form-fn))
    (funcall body-fn forms)))

(defmacro with-collected-macro-forms ((var) instrumented-form &body body)
  "Collect macro forms by locally binding *MACROEXPAND-HOOK*.

Evaluates INSTRUMENTED-FORM, with *MACROEXPAND-HOOK* bound to a
function which collects all forms that undergo macro-expansion into a
list stored in VAR.  Then evaluates BODY, with VAR still bound to the
list of forms."
  (assert (symbolp var))
  (assert (not (null var)))
  `(call-with-collected-macro-forms (lambda (,var) ,@body)
				    (lambda () ,instrumented-form)))```